make torch._check understand Eq commutativity

avikchaudhuri opened the pull request:
Summary:
Given `torch._check(a == b)` we can still get a data-dependent error needing `b == a`. Simple fix.

```
def forward(self, x1, x2, x3, y):
    z1 = x1.item()
    z2 = x2.item()
    z3 = x3.item()
    torch._check((z2 + z3) == z1)
    # torch._check(z1 == (z2 + z3)) didn't work, now does
    if z2 + z3 == z1:
        return y * 2
    else:
        return y + 3
```

Differential Revision: D57014730

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -4017,6 +4017,7 @@ def add_expr(expr):
         add_expr(e)
         # Other relational expressions this expression implies
         if isinstance(e, sympy.Eq):
+            add_expr(sympy.Eq(e.rhs, e.lhs))
```
btw, also noticed that the dualizing above didn't have an effect, I think canonicalizing after dualizing cancels out. maybe something to follow up in the future

avikchaudhuri replied:
for the code above, we get something like `Eq(u0 - u1 - u2, 0)` to evaluate statically with substitutions like `Eq(u1 + u2 - u0, 0)`, which didn't pattern match.

lezcano replied:
Ugh, right. To kill two birds with one stone, what you can do is, in the case of Eq, take the dual of `a = b` as `-a = -b`. The canonicalisation of this expression would give you `b - a = 0`.

ezyang reviewed (approved):
Test please.

@lezcano do you want this as is, or would you prefer we land the two birds directly?

lezcano commented:
Probably the proposed change, as it's also one line

avikchaudhuri committed:
make torch._check understand Eq commutativity (#125629)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/125629

Given `torch._check(a == b)` we can still get a data-dependent error needing `b == a`. Simple fix.

```
def forward(self, x1, x2, x3, y):
    z1 = x1.item()
    z2 = x2.item()
    z3 = x3.item()
    torch._check((z2 + z3) == z1)
    # torch._check(z1 == (z2 + z3)) didn't work, now does
    if z2 + z3 == z1:
        return y * 2
    else:
        return y + 3
```

Test Plan: added test

Reviewed By: ezyang

Differential Revision: D57014730
